internal/provider/user_role_grant_resource.go

```diff
@@ -0,0 +1,272 @@
+
+	for _, respRole := range respUserRoles {
+		if roleEqualsBuiltInRole(&state.Role, &respRole) {
+			// If we did find the matching role grant there isn't anything
+			// else to do because these roles are immutable.
+			return
+		}
+	}
+
+	// If we never found the matching role grant we need to remove
+	// it from the state.
+	if len(respUserRoles) == 0 {
+		resp.Diagnostics.AddWarning(
+			"Couldn't find user role grant.",
+			fmt.Sprintf(
+				"Could not find any roles for a user matching the id=%s. Removing role grant from state.",
+				stateUserID,
+			),
+		)
+	} else {
+		resp.Diagnostics.AddWarning(
+			"Couldn't find user role grant.",
+			fmt.Sprintf(
+				"Could not find user role grant for user (id=%s) and role (%s/%s/%s). Removing role grant from state.",
+				stateUserID,
+				state.Role.RoleName.ValueString(),
+				state.Role.ResourceType.ValueString(),
+				state.Role.ResourceId.ValueString(),
+			),
+		)
+	}
+	resp.State.RemoveResource(ctx)
+}
+
+// This resource is immutable so no action is necessary
+func (r *userRoleGrantResource) Update(
+	ctx context.Context, req resource.UpdateRequest, resp *resource.UpdateResponse,
+) {
+}
+
+func (r *userRoleGrantResource) Delete(
+	ctx context.Context, req resource.DeleteRequest, resp *resource.DeleteResponse,
+) {
+	var state UserRoleGrant
+	diags := req.State.Get(ctx, &state)
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	_, _, err := r.provider.service.RemoveUserFromRole(ctx, state.UserID.ValueString(), state.Role.ResourceType.ValueString(), state.Role.ResourceId.ValueString(), state.Role.RoleName.ValueString())
+	if err != nil {
+		resp.Diagnostics.AddError(
+			"Error remove user role grant for user",
+			fmt.Sprintf("Could not remove user role grant for user: %s", formatAPIErrorMessage(err)),
+		)
+		return
+	}
+
+	resp.State.RemoveResource(ctx)
+}
+
+func (r *userRoleGrantResource) ImportState(
+	ctx context.Context, req resource.ImportStateRequest, resp *resource.ImportStateResponse,
+) {
+	userID, roleName, resourceType, resourceID, err := parseRoleGrantResourceID(req.ID)
+	if err != nil {
+		resp.Diagnostics.AddError(
+			"Unexpected Import Identifier",
+			err.Error(),
+		)
+		return
+	}
+
+	role := Role{
+		RoleName:     types.StringValue(roleName),
+		ResourceType: types.StringValue(resourceType),
+		ResourceId:   types.StringValue(resourceID),
+	}
+
+	resp.Diagnostics.Append(resp.State.SetAttribute(ctx, path.Root("user_id"), userID)...)
+	resp.Diagnostics.Append(resp.State.SetAttribute(ctx, path.Root("role"), role)...)
+}
+
+func NewUserRoleGrantResource() resource.Resource {
+	return &userRoleGrantResource{}
+}
+
+const resourceTypeOrg = string(client.RESOURCETYPETYPE_ORGANIZATION)
+
+func parseRoleGrantResourceID(
+	id string,
+) (userID, roleName, resourceType, resourceID string, err error) {
+	idParts := strings.Split(id, ",")
+	if len(idParts) == 4 {
+		if idParts[2] == resourceTypeOrg && idParts[3] != "" {
```

Comment:
Maybe a `uuid.Parse` would be helpful here as a sanity check

Reply:
I've updated that method and tests to do checking on the uuids.